src/coap_oscore.c

```diff
@@ -0,0 +1,2100 @@
+                                               sizeof(external_aad_buffer));
+    cose_encrypt0_set_external_aad(cose, &external_aad);
+
+    /* AAD */
+    aad.s = aad_buffer;
+    aad.length = oscore_prepare_aad(external_aad_buffer,
+                                    external_aad.length,
+                                    aad_buffer,
+                                    sizeof(aad_buffer));
+    assert(aad.length < AAD_BUF_LEN);
+    cose_encrypt0_set_aad(cose, &aad);
+
+    /*
+     * RFC8613 8.2 Step 5.
+     * Compute the AEAD nonce.
+     *
+     * Requires in COSE object as appropriate
+     *   key_id (kid) (Recipient ID)
+     *   partial_iv   (as received in request)
+     *   common_iv    (already in osc_ctx)
+     */
+    nonce.s = nonce_buffer;
+    nonce.length = 13;
+    oscore_generate_nonce(cose, osc_ctx, nonce_buffer, 13);
+    cose_encrypt0_set_nonce(cose, &nonce);
+    /*
+     * Set up an association for use in the response
+     */
+    association = oscore_find_association(session, &pdu_token);
+    if (association) {
+      /* Refresh the association */
+      coap_delete_bin_const(association->nonce);
+      association->nonce =
+          coap_new_bin_const(cose->nonce.s, cose->nonce.length);
+      if (association->nonce == NULL)
+        goto error;
+      coap_delete_bin_const(association->partial_iv);
+      association->partial_iv =
+          coap_new_bin_const(cose->partial_iv.s, cose->partial_iv.length);
+      if (association->partial_iv == NULL)
+        goto error;
+      coap_delete_bin_const(association->aad);
+      association->aad = coap_new_bin_const(cose->aad.s, cose->aad.length);
+      if (association->aad == NULL)
+        goto error;
+      association->recipient_ctx = rcp_ctx;
+    } else if (!oscore_new_association(session,
+                                       NULL,
+                                       &pdu_token,
+                                       rcp_ctx,
+                                       &cose->aad,
+                                       &cose->nonce,
+                                       &cose->partial_iv,
+                                       0)) {
+      goto error;
+    }
+    /* So association is not released when handling decrypt */
+    association = NULL;
+  } else /* ! coap_request */ {
+    /* Need to do nonce before AAD because of different partial_iv */
+    /*
+     * 8.4 Step 4.
+     * Compose the AEAD nonce.
+     */
+    cose_encrypt0_set_key_id(cose, rcp_ctx->recipient_id);
+    if (cose->partial_iv.length == 0) {
+      cose_encrypt0_set_partial_iv(cose, association->partial_iv);
+      cose_encrypt0_set_nonce(cose, association->nonce);
+    } else {
+      /*
+       * Requires in COSE object as appropriate
+       *   kid (set above)
+       *   partial_iv (as received)
+       *   common_iv (already in osc_ctx)
+       */
+      oscore_generate_nonce(cose, osc_ctx, nonce_buffer, 13);
```

Comment:
This looks dangerous. In the case that a sequence number is contained, there appears to be no check whether this sequence number actually matches the one used in the corresponding request. An attacker might take an older response, insert a new Message ID and Token, and get this replayed response accepted.

Reply:
Good thinking. I will look for a way to verify that the sequence number is as expected.